### PR TITLE
Shop test

### PR DESCRIPTION
**Pull request by chuck-dbos**

### CODE REVIEW

File: e-commerce/payment-backend/src/operations.ts [hunk 1]
```diff
@@ -175,7 +175,11 @@ export class PlaidPayments {
     session_id: string, 
     status: string | undefined, 
     client_reference_id: string | undefined
-  ): Promise<void> {
+  ): Promise<void>
+  {
+    if (webhook === "http://fakehost/webhook") {
```
Comment (qianl15) [inline e-commerce/payment-backend/src/operations.ts:180]: In the operations.test.ts, can we try to use `spyOn` for the `fetch` call?

Comment (qianl15) [reply e-commerce/payment-backend/src/operations.ts:180]: Like this:
```
const fetchMock = jest
  .spyOn(global, 'fetch')
  .mockImplementation(() =>
    Promise.resolve({ json: () => Promise.resolve([]) })
  )
```
(https://stackoverflow.com/a/73597163)

File: e-commerce/shop-backend/src/operations.test.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    */
+
+    // Initiate checkout
+    const handle = await testRuntime.invoke(Shop).paymentWorkflow('shopper', 'xxx');
+    const url = await testRuntime.getEvent<string>(handle.getWorkflowUUID(), checkout_url_topic);
+    if (!url) throw new Error("URL not returned");
+
+    // Fake a payment reply 
+    const payresp = await request(testRuntime.getHandlersCallback())
+    .post(`/payment_webhook`).send({session_id: "1234", client_reference_id: url.replace(/^[^>]*>/g, '').replace(/<.*/g, ''), payment_status: "paid"});
```
Comment (qianl15) [inline e-commerce/shop-backend/src/operations.test.ts:103]: What does this part do? `url.replace(/^[^>]*>/g, '').replace(/<.*/g, '')`

File: e-commerce/shop-backend/src/operations.ts [hunk 3]
```diff
@@ -264,6 +267,14 @@ export class Shop {
...
+        // TODO: Only in test mode
+        return {
+            session_id: "1234",
+            url:ctxt.workflowUUID,
+            payment_status: "pending",
+        };
+    }
     const { paymentHost, localHost } = getHostConfig(ctxt);
 
     const response = await fetch(`${paymentHost}/api/create_payment_session`, {
```
Comment (qianl15) [inline e-commerce/shop-backend/src/operations.ts:280]: If we use `spyOn` for this `fetch` call, maybe we don't need to do the `origin == 'xxx'` trick above.

Comment (chuck-dbos) [reply e-commerce/shop-backend/src/operations.ts:280]: I have a better idea, I think.

File: e-commerce/payment-backend/src/operations.ts (hunk 1 above)
Comment (chuck-dbos) [reply e-commerce/payment-backend/src/operations.ts:180]: I was going to illustrate both ways.  I think it is fair to have some provision for behaviors that occur only in test mode, as mocks can get very complicated and almost as hard to debug as the real code.

File: e-commerce/shop-backend/src/operations.test.ts (hunk 2 above)
Comment (chuck-dbos) [reply e-commerce/shop-backend/src/operations.test.ts]: Originally I was working with URLs only.  Now that I invoke the workflow (not using request handler) I can get the workflow ID to pass to the payment callback handler.

However, the old code did have a test for the redirect... I sort of lost the test of the redirect.

File: e-commerce/shop-backend/src/operations.ts (hunk 3 above)
Comment (chuck-dbos) [reply e-commerce/shop-backend/src/operations.ts:276]: It is unfortunate that spyOn doesn't work in a straightforward manner on our code.  I worked around it in a less brutal way than stubbing out fetch entirely... and in a way that would be illustrative of more complex scenarios.

**Review by qianl15 — APPROVED** [review]

File: e-commerce/payment-backend/src/operations.ts (hunk 1 above)
Comment (qianl15) [reply e-commerce/payment-backend/src/operations.ts:180]: Makes sense

**Pull request merged**